### PR TITLE
Prompt for email verification when using GDevelop

### PR DESCRIPTION
**Pull request by ClementPasteau**

### CODE REVIEW

File: newIDE/app/src/Profile/AuthenticatedUserContext.js [hunk 1]
```diff
@@ -40,6 +40,10 @@ export type AuthenticatedUser = {|
   onSubscriptionUpdated: () => Promise<void>,
   onPurchaseSuccessful: () => Promise<void>,
   onSendEmailVerification: () => Promise<void>,
+  onOpenEmailVerificationDialog: ({
+    sendEmailAutomatically: boolean,
+    showSendEmailButton: boolean,
+  }) => void,
```
Comment (AlexandreSi) [inline newIDE/app/src/Profile/AuthenticatedUserContext.js:46]: Use exact type `{| |}`

File: newIDE/app/src/Profile/AuthenticatedUserProvider.js [hunk 2]
```diff
@@ -58,7 +58,11 @@ type State = {|
   additionalUserInfoDialogOpen: boolean,
   authError: ?AuthError,
   resetPasswordDialogOpen: boolean,
-  emailVerificationPendingDialogOpen: boolean,
+  emailVerificationDialogOpen: boolean,
+  emailVerificationDialogProps: {
```
Comment (AlexandreSi) [inline newIDE/app/src/Profile/AuthenticatedUserProvider.js:62]: Use exact type here as well

File: newIDE/app/src/Profile/AuthenticatedUserProvider.js [hunk 3]
```diff
@@ -152,12 +161,25 @@ export default class AuthenticatedUserProvider extends React.Component<
         onSubscriptionUpdated: this._fetchUserSubscriptionLimitsAndUsages,
         onPurchaseSuccessful: this._fetchUserAssets,
         onSendEmailVerification: this._doSendEmailVerification,
+        onOpenEmailVerificationDialog: ({
+          sendEmailAutomatically,
+          showSendEmailButton,
+        }: {
```
Comment (AlexandreSi) [inline newIDE/app/src/Profile/AuthenticatedUserProvider.js:167]: Ditto

File: newIDE/app/src/stories/index.js [hunk 4]
```diff
@@ -2708,22 +2706,6 @@ storiesOf('AuthenticatedUserProfileDetails', module)
     />
   ));
 
-storiesOf('EmailVerificationPendingDialog', module)
```
Comment (AlexandreSi) [inline newIDE/app/src/stories/index.js:2711]: 💜

File: newIDE/app/src/Utils/GDevelopServices/Authentication.js [hunk 5]
```diff
@@ -220,16 +223,14 @@ export default class Authentication {
     const { currentUser } = this.auth;
     if (!currentUser || currentUser.emailVerified) return;
 
-    try {
-      sendEmailVerification(currentUser);
-    } catch (error) {
+    sendEmailVerification(currentUser).catch(error => {
```
Comment (AlexandreSi) [inline newIDE/app/src/Utils/GDevelopServices/Authentication.js:226]: Why this change?

Comment (AlexandreSi) [reply newIDE/app/src/Utils/GDevelopServices/Authentication.js:226]: We usually do the other way around change: promise to async/await

File: newIDE/app/src/Profile/EmailVerificationDialog.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
+// @flow
+import { Trans } from '@lingui/macro';
+
+import React from 'react';
+import FlatButton from '../UI/FlatButton';
+import Dialog, { DialogPrimaryButton } from '../UI/Dialog';
+import { type AuthenticatedUser } from './AuthenticatedUserContext';
+import BackgroundText from '../UI/BackgroundText';
+import VerifiedUser from '@material-ui/icons/VerifiedUser';
```
Comment (AlexandreSi) [inline newIDE/app/src/Profile/EmailVerificationDialog.js:9]: You could use this icon from our new icons:
image

File: newIDE/app/src/Profile/EmailVerificationDialog.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                  onSendEmail();
+                }}
+              />
+            ) : (
+              <BackgroundText>
+                <LineStackLayout justifyContent="center" alignItems="center">
+                  <CircularProgress size={20} />
+                  <Text>
+                    <Trans>
+                      Waiting for the email verification of
```
Comment (AlexandreSi) [inline newIDE/app/src/Profile/EmailVerificationDialog.js:117]: I don't know if the english is right here, "A verification email has been sent to {}, waiting for validation". But I'm not sure it's better

File: newIDE/app/src/Profile/AuthenticatedUserProvider.js [hunk 8]
```diff
@@ -755,10 +812,28 @@ export default class AuthenticatedUserProvider extends React.Component<
     this._automaticallyUpdateUserProfile = true;
   };
 
-  openEmailVerificationPendingDialog = (open: boolean = true) => {
+  openEmailVerificationDialog = ({
+    open = true,
+    sendEmailAutomatically = false,
+    showSendEmailButton = false,
+  }: {
```
Comment (AlexandreSi) [inline newIDE/app/src/Profile/AuthenticatedUserProvider.js:819]: Use exact type here as well

File: newIDE/app/src/Profile/AuthenticatedUserProvider.js [hunk 9]
```diff
@@ -528,6 +541,51 @@ export default class AuthenticatedUserProvider extends React.Component<
...
+      this._notifyUserAboutAdditionalInfo();
+    }
+  };
+
+  _notifyUserAboutAdditionalInfo = () => {
+    const profile = this.state.authenticatedUser.profile;
+    if (!profile) return;
+    // If the user has not filled their additional information, show
+    // the dialog to fill it, but ensure they have closed the email verification dialog first.
+    // Use a state value to show the dialog only once.
```
Comment (AlexandreSi) [inline newIDE/app/src/Profile/AuthenticatedUserProvider.js:579]: That's not really a state value, it it?

File: newIDE/app/src/Profile/AuthenticatedUserProvider.js [hunk 10]
```diff
@@ -528,6 +541,51 @@ export default class AuthenticatedUserProvider extends React.Component<
...
   };
 
+  _notifyUserAboutEmailVerificationAndAdditionalInfo = () => {
+    const { profile } = this.state.authenticatedUser;
+    if (!profile) return;
+    // If the user has not verified their email when logging in we show a dialog to do so.
+    // - If they just registered, we don't send the email again as it will be sent automatically,
+    // nor do we show a button to send again.
+    // - If they are just logging in, we don't send the email but we show a button to send again.
+    // Use a state value to show the dialog only once.
```
Comment (AlexandreSi) [inline newIDE/app/src/Profile/AuthenticatedUserProvider.js:551]: Same here for the state value mention

File: newIDE/app/src/Profile/AuthenticatedUserProvider.js [hunk 11]
```diff
@@ -875,17 +950,26 @@ export default class AuthenticatedUserProvider extends React.Component<
...
                     .onRefreshFirebaseProfile()
                     .catch(() => {
                       // Ignore any error, we can't do much.
                     });
                 }}
+                sendEmailAutomatically={
+                  this.state.emailVerificationDialogProps.sendEmailAutomatically
+                }
+                showSendEmailButton={
+                  this.state.emailVerificationDialogProps.showSendEmailButton
```
Comment (AlexandreSi) [inline newIDE/app/src/Profile/AuthenticatedUserProvider.js:970]: Should you just spread the props with `{...this.state.emailVerificationDialogProps}`?

File: newIDE/app/src/Profile/EmailVerificationDialog.js (hunk 6 above)
Comment (ClementPasteau) [reply newIDE/app/src/Profile/EmailVerificationDialog.js:9]: I'll use this one

File: newIDE/app/src/Profile/AuthenticatedUserProvider.js (hunk 9 above)
Comment (ClementPasteau) [reply newIDE/app/src/Profile/AuthenticatedUserProvider.js:579]: kinda is, using this.xxx use the React state internally I believe. I'll update the wording if that's confusing

**Review by AlexandreSi — APPROVED** [review]
LGTM
